### PR TITLE
Conversion from Google Calendar Event to Event

### PR DESCRIPTION
**Pull request by Flechman**
Please pay special attention to the logic to translate the recurrence, tell me what you would do better.

### CODE REVIEW

File: app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java [hunk 1]
```diff
@@ -19,6 +25,7 @@
     private static final String OAUTH_SCOPE = "oauth2:profile email";
     private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
     static final String LOGCAT_TAG = "GCalendar";
+    private static final String DEFAULT_NAME = "No Name";
```
Comment (hugolan) [inline app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:28]: why would the default name be no name? maybe put something generic there

**Review by hugolan — APPROVED** [review]
good code nothing much to say here

File: app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java (hunk 1 above)
Comment (Flechman) [reply app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:38]: What are your suggestions ? "No name" is good as it gives a hint to the user to give a real event name.

Comment (hugolan) [reply app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:40]: I was thinking about something like insert name but ok

**Review by hugolan — APPROVED** [review]

File: app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java [hunk 2]
```diff
@@ -111,10 +107,67 @@ public static Event toGCalendarEvents(com.github.onedirection.event.Event event)
...
+    public static Event fromGCalendarEvents(com.google.api.services.calendar.model.Event event) {
+        Objects.requireNonNull(event);
+        if(event.getStart() == null || event.getEnd() == null) {
+            throw new IllegalArgumentException("Event should have at least a start time and an end time.");
+        }
+
+        Id newId = Id.generateRandom();
+        String name = (event.getSummary() != null) ? event.getSummary() : DEFAULT_NAME;
+        //TODO: ask @Ef55 what default location name to put if no location is specified
+        String locationName = (event.getLocation() != null) ? event.getLocation() : "";
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:128]: I think that an empty string will do just fine.

File: app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java [hunk 3]
```diff
@@ -111,10 +107,67 @@ public static Event toGCalendarEvents(com.github.onedirection.event.Event event)
...
+        if(event.getRecurrence() != null) {
+            List<String> recurrences = event.getRecurrence();
+            String periodically = null;
+            int eventCount = 0;
+
+            for(String rule : recurrences) {
+                if(rule.substring(0, 5).equals("RRULE")) {
+                    String[] info = rule.substring(11).split(";COUNT=");
+                    periodically = info[0];
+                    eventCount = Integer.parseInt(info[1]);
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:146]: - I think a little link to the doc wouldn't hurt (to answer questions like "why 11 ?" or "can info.length > 2 ?");
- I guess that no exception is thrown if multiple rules match, which might not be good.

File: app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java [hunk 4]
```diff
@@ -111,10 +107,67 @@ public static Event toGCalendarEvents(com.github.onedirection.event.Event event)
...
+            if(periodically == null) {
+                throw new IllegalArgumentException("No recurrence rule matches the one used by "+ R.string.app_name+" Events");
+            }
+
+            Duration period = null;
+            for(Map.Entry<TemporalUnit, String> t : PERIODS.entrySet()) {
+                if(periodically.equals(t.getValue())) {
+                    period = t.getKey().getDuration();
+                }
+            }
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:158]: I think that this map could be put to better use:

- To do _TemporalUnit_ -> _String_, I guess that _.toString()_ might be sufficient;
- The key/values could then be reversed, and you could then simply do `PERIODS.get(periodically).getDuration()`.

**Review by Ef55 — CHANGES_REQUESTED** [review]
Just a few details to fix, the overall logic seems good.

I would also strongly encourage you to add a "coherence test", where you convert an event into its GCalendar form, and then convert it back into our Event, to final test equality with the original event.

**Comment by Flechman** [comment]
> Just a few details to fix, the overall logic seems good.
> 
> I would also strongly encourage you to add a "coherence test", where you convert an event into its GCalendar form, and then convert it back into our Event, to final test equality with the original event.

My test on the conversion from GCalendar to Event does this check already. I cannot test for equality on events since the conversion changes the ids (because GCalendar id formatting may not be the same as UUID).

File: app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java (hunk 3 above)
Comment (Flechman) [reply app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:146]: Done

File: app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java (hunk 4 above)
Comment (Flechman) [reply app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:158]: I don't understand your first point, what is it for ? 
Note that the String is a period such as "DAILY", not a  technical duration format

Comment (Flechman) [reply app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java:158]: Second point is done

Comment (Ef55) [reply app/src/main/java/com/github/onedirection/interoperability/gcalendar/GoogleCalendar.java]: You're indeed right; didn't realized that the strings weren't simply the name of the period.

**Comment by Ef55** [comment]
> My test on the conversion from GCalendar to Event does this check already. I cannot test for equality on events since the conversion changes the ids (because GCalendar id formatting may not be the same as UUID).

This might be slightly annoying when we will import events back from GCalendar. 
We will have to discuss that with the «Product owner», and do changes (in a subsequent PR) accordingly.

**Review by Ef55 — APPROVED** [review]

**Pull request merged**